### ISSUE TITLE
Adding a devfile to devfile registry at runtime doesn't work

### ISSUE THREAD

**Issue by VeerMuchandi** [eclipse-che/che#14691]
### Describe the bug
I am following instructions here https://www.eclipse.org/che/docs/che-7/editing-a-devfile-and-plug-in-at-runtime/#adding-a-devfile-at-runtime_editing-a-devfile-and-plug-in-at-runtime

This step fails
```
$ kubectl exec ${DEVFILE_REG_POD} -i -t -- \
  /var/www/html/index.sh > /var/www/html/devfiles/index.json
```

with error `-bash: /var/www/html/devfiles/index.json: No such file or directory`

Also there are errors with `\` in `$\{DEVFILE_REG_POD} ` throughout the documentation. I had to remove that everywhere.

I tried to do this by rshing into the pod

```
$ oc rsh ${DEVFILE_REG_POD} 
sh-4.2$ /var/www/html/index.sh > /var/www/html/devfiles/index.json
/var/www/html/index.sh: line 19: readarray: -d: invalid option
readarray: usage: readarray [-n count] [-O origin] [-s count] [-t] [-u fd] [-C callback] [-c quantum] [array]
sh-4.2$ find: 'devfiles': No such file or directory
```
The actual error is the `-d` option in the script is not recognized.

### Che version

 - [ ] latest
 - [ ] nightly
 - [ ] other: please specify

### Steps to reproduce

Add a devfile to devfile registry as explained in the documentation to test

### Expected behavior
`index.json` should have new entries and the stack should show up on the screen

### Runtime
   OCP 4.1.18
  
```
$ oc version
Client Version: version.Info{Major:"4", Minor:"1+", GitVersion:"v4.1.7-201907171753+dc76278-dirty", GitCommit:"dc76278", GitTreeState:"dirty", BuildDate:"2019-07-18T00:52:21Z", GoVersion:"go1.11.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.4+c2a5caf", GitCommit:"c2a5caf", GitTreeState:"clean", BuildDate:"2019-09-21T02:12:52Z", GoVersion:"go1.11.13", Compiler:"gc", Platform:"linux/amd64"}
```
### Installation method
   - [ ] chectl 
   - [ x] che-operator 
   - [ ] minishift-addon
   - [ ] I don't know

### Environment
   - [ ] my computer 
       - [ ] Windows
       - [ ] Linux
       - [x ] macOS
   - [ ] Cloud
       - [ ] Amazon
       - [ ] Azure
       - [ ] GCE
       - [ ] other (please specify) 
   - [ ] other: please specify

**Comment by tolusha** [comment]
I marked this issue as `P1` since it is very important to customize plugin and dev registries

**Comment by tsmaeder** [comment]
Folks, this is about the process to add devfiles to the registry on the fly by hand. IMO, this process is a hack should not even be in user documentation. The additions will go away the next time the registry pod is restarted. 
The right thing to do is to fork the registry.

**Comment by amisevsk** [comment]
I agree with @tsmaeder, I see no reason why you would want to add your devfile to a running registry, when there are so many better options already:
- Pastebin/gist with factory URL
- Github repo with devfile checked in
- Fork devfile registry if you consistently need it.

**Comment by tylerauerbeck** [comment]
I think the runtime concern is valid, you probably don't want to have to do this everytime the pod restarts. That being said, it may be useful to still include the index.sh script so that instead of having to maintain a seperate fork of the repo, I can just say FROM quay.io/eclipse/che-devfile-registry:nightly... and then add my custom definitions that way.

**Comment by gorkem** [comment]
This also come up lately in the context of the odo adoption of devfiles. I do not think we can continue to recommend the current _hackish_ form of updating the registry. @neeraj-laad may have some ideas that was used on appsody.  I think we need to provide a tool like [`appsody stack`](https://appsody.dev/docs/cli-commands/#appsody-stack) command for registry management.

**Comment by themr0c** [comment]
Should we completely remove https://www.eclipse.org/che/docs/che-7/end-user-guide/editing-a-devfile-and-plug-in-at-runtime/ then?

**Comment by themr0c** [comment]
Referenced in https://issues.redhat.com/browse/RHDEVDOCS-2278

**Comment by KOSSOKO** [comment]
@tolusha, So what is the best solution to do this ? Because we are trying to add our custom devfile in the existing devfile-registry without success. The command still fails.

**Comment by amisevsk** [comment]
@KOSSOKO Probably the easiest fix is to update `bash` -- version 4.4 introduced the `-d` option for `readarray`. 

However, if you're using the dockerimage to build the registry, you shouldn't be seeing this error.

**Comment by KOSSOKO** [comment]
@amisevsk yes, i noticed that the with the version of bash i have,` -d` is an undefined option. So based ok the docker image of the devfile registry we rebuild our own image registry and our devfiles inside. And we are trying to use this docker image during chectl install with helm.

**Comment by amisevsk** [comment]
@KOSSOKO I'm not sure I understand your process -- are you basing your custom devfile registry off of `quay.io/eclipse/che-devfile-registry`? I'm not sure where the old version of bash is entering the equation.

The simplest way to build your own registry is to use the provided `build.sh` script in the `che-devfile-registry` repo (see `./build.sh --help` for supported flags):
```bash
git clone https://github.com/eclipse/che-devfile-registry.git
cd che-devfile-registry
# Add your devfile and a meta.yaml to ./devfiles/your-organization/
./build.sh \
    --registry docker.io # or quay.io, or whichever registry you're pushing to
    --organization my-docker-organization \
    --tag my-tag
```
This will build the usual devfile registry in image `docker.io/my-docker-organization/che-plugin-registry:my-tag`

**Comment by KOSSOKO** [comment]
@amisevsk yes, we tried the kubernetes version, but we have 2 registries at he end.  The first deployed by chectl and the registry build by the command build.sh. We could not have two registries.

So we decided to fork the registry project on github,  add our devfiles, update the index.json, and rebuild the docker image and hosted it in our cloud docker images.

During chectl intall, we tell to chectl to use a dedicaded helm file for the install.
As result, we have one registry and all images provided by default by che and our devfiles.

**Comment by ibuziuk** [comment]
Closing since https://github.com/eclipse-che/che-devfile-registry was archived in favour of https://registry.devfile.io/
